File: educational-core/src/com/jetbrains/edu/learning/EduCourseBuilder.java
package com.jetbrains.edu.learning;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.jetbrains.edu.coursecreator.actions.CCCreateLesson;
import com.jetbrains.edu.coursecreator.actions.CCCreateTask;
import com.jetbrains.edu.coursecreator.actions.NewStudyItemInfo;
import com.jetbrains.edu.coursecreator.actions.NewStudyItemUiModel;
import com.jetbrains.edu.coursecreator.ui.AdditionalPanel;
import com.jetbrains.edu.coursecreator.ui.NewStudyItemUiUtils;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseFormat.Lesson;
import com.jetbrains.edu.learning.courseFormat.TaskFile;
import com.jetbrains.edu.learning.courseFormat.ext.TaskExt;
import com.jetbrains.edu.learning.courseFormat.tasks.Task;
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils;
import com.jetbrains.edu.learning.newproject.CourseProjectGenerator;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.List;

/**
 * The main interface provides courses creation for some language.
 *
 * @param <Settings> container type holds course project settings state
 */
public interface EduCourseBuilder<Settings> {

  Logger LOG = Logger.getInstance(EduCourseBuilder.class);

  /**
   * Shows UI for new study item creation
   *
   * @param model some parameters for UI extracted from context where creating action was called
   * @param additionalPanels additional ui elements which should be shown while new study item creation
   *
   * @see com.jetbrains.edu.coursecreator.ui.CCItemPositionPanel
   *
   * @return properties for study item creation
   */
  @Nullable
  default NewStudyItemInfo showNewStudyItemUi(@NotNull Project project,
                                              @NotNull NewStudyItemUiModel model,
                                              @NotNull List<AdditionalPanel> additionalPanels) {
    return NewStudyItemUiUtils.showNewStudyItemDialog(project, model, additionalPanels);
  }

  /**
   * Creates content (including its directory or module) of new lesson in project
   *
   * @param project Parameter is used in Java and Kotlin plugins
   * @param lesson  Lesson to create content for. It's already properly initialized and added to course.
   * @return VirtualFile of created lesson
   */
  default VirtualFile createLessonContent(@NotNull Project project,
                                          @NotNull Lesson lesson,
                                          @NotNull VirtualFile parentDirectory) {
    final VirtualFile[] lessonDirectory = new VirtualFile[1];
    ApplicationManager.getApplication().runWriteAction(() -> {
      try {
        lessonDirectory[0] = VfsUtil.createDirectoryIfMissing(parentDirectory, lesson.getName());
      } catch (IOException e) {
        LOG.error("Failed to create lesson directory", e);
      }
    });
    return lessonDirectory[0];
  }

  /**
   * Creates content (including its directory or module) of new task in project
   *
   * @param task Task to create content for. It's already properly initialized and added to corresponding lesson.
   * @return VirtualFile of created task
   */
  @Nullable
  default VirtualFile createTaskContent(@NotNull final Project project,
                                        @NotNull final Task task,
                                        @NotNull final VirtualFile parentDirectory) {
    try {
      GeneratorUtils.createTask(task, parentDirectory);
    } catch (IOException e) {
      LOG.error("Failed to create task", e);
    }
    final VirtualFile taskDir = parentDirectory.findChild(task.getName());
    if (!OpenApiExtKt.isUnitTestMode()) {
      refreshProject(project);
    }
    return taskDir;
  }

  /**
   * Allows to update project modules and the whole project structure
   */
  default void refreshProject(@NotNull final Project project) {
    refreshProject(project, null);
  }

  default void refreshProject(@NotNull final Project project, @Nullable ProjectRefreshListener listener) {
    if (listener != null) {
      listener.onSuccess();
    }
  }

  @Nullable
  default Lesson createInitialLesson(@NotNull Project project, @NotNull Course course) {
    Lesson lesson = new CCCreateLesson().createAndInitItem(project, course, null, new NewStudyItemInfo(EduNames.LESSON + 1, 1));
    Task task = new CCCreateTask().createAndInitItem(project, course, lesson, new NewStudyItemInfo(EduNames.TASK + 1, 1));
    if (task != null) {
      lesson.addTask(task);
    }
    return lesson;
  }

  /**
   * Add initial content for a new task: task and tests files if the corresponding files don't exist.
   * Supposed to use in course creator mode
   *
   * @param task initializing task
   */
  default void initNewTask(@NotNull Project project, @NotNull final Lesson lesson, @NotNull final Task task, @NotNull NewStudyItemInfo info) {
    if (task.getTaskFiles().isEmpty()) {
      String sourceDir = TaskExt.getSourceDir(task);
      TaskFile taskFile = new TaskFile();
      String taskTemplateName = getTaskTemplateName();
      if (taskTemplateName != null) {
        taskFile.setName(GeneratorUtils.joinPaths(sourceDir, taskTemplateName));
        taskFile.setText(StringUtil.notNullize(GeneratorUtils.getInternalTemplateText(taskTemplateName)));
      } else {
        GeneratorUtils.DefaultFileProperties taskFileProperties =
          GeneratorUtils.createDefaultFile(task.getLesson().getCourse(), "Task", "type task text here");
        taskFile.setName(GeneratorUtils.joinPaths(sourceDir, taskFileProperties.getName()));
        taskFile.setText(taskFileProperties.getText());
      }
      task.addTaskFile(taskFile);
      TaskFile defaultTestFile = createDefaultTestFile(task);
      if (defaultTestFile != null) {
        task.addTaskFile(defaultTestFile);
      }
    }
  }

  @Nullable
  default TaskFile createDefaultTestFile(@NotNull Task task) {
    List<String> testDirs = TaskExt.getTestDirs(task);
    String testDir = "";
    if (!testDirs.isEmpty()) {
      testDir = testDirs.get(0);
    }
    String testTemplateName = getTestTemplateName();
    if (testTemplateName != null) {
      String testText = GeneratorUtils.getInternalTemplateText(testTemplateName);
      TaskFile test =
        new TaskFile(GeneratorUtils.joinPaths(testDir, testTemplateName), testText);
      test.setVisible(false);
      return test;
    }
    return null;
  }

  @Nullable
  default String getTaskTemplateName() {
    return null;
  }

  @Nullable
  default String getTestTemplateName() {
    return null;
  }

  /**
   * @return object responsible for language settings
   * @see LanguageSettings
   */
  @NotNull
  LanguageSettings<Settings> getLanguageSettings();

  @Nullable
  default CourseProjectGenerator<Settings> getCourseProjectGenerator(@NotNull Course course) {
    return null;
  }

  interface ProjectRefreshListener {
    default void onSuccess() {}
    default void onFailure(@NotNull String errorMessage) {}
  }
}

File: educational-core/src/com/jetbrains/edu/learning/EduCourseBuilder.kt
package com.jetbrains.edu.learning

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.jetbrains.edu.coursecreator.StudyItemType
import com.jetbrains.edu.coursecreator.actions.TemplateFileInfo
import com.jetbrains.edu.coursecreator.actions.studyItem.CCCreateLesson
import com.jetbrains.edu.coursecreator.actions.studyItem.CCCreateTask
import com.jetbrains.edu.coursecreator.actions.studyItem.NewStudyItemInfo
import com.jetbrains.edu.coursecreator.actions.studyItem.NewStudyItemUiModel
import com.jetbrains.edu.coursecreator.ui.showNewStudyItemDialog
import com.jetbrains.edu.learning.courseFormat.*
import com.jetbrains.edu.learning.courseFormat.EduFormatNames.LESSON
import com.jetbrains.edu.learning.courseFormat.EduFormatNames.TASK
import com.jetbrains.edu.learning.courseFormat.ext.configurator
import com.jetbrains.edu.learning.courseFormat.ext.testDirs
import com.jetbrains.edu.learning.courseFormat.tasks.*
import com.jetbrains.edu.learning.courseFormat.tasks.OutputTaskBase.Companion.INPUT_PATTERN_NAME
import com.jetbrains.edu.learning.courseFormat.tasks.OutputTaskBase.Companion.OUTPUT_PATTERN_NAME
import com.jetbrains.edu.learning.courseFormat.tasks.choice.ChoiceTask
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils.joinPaths
import com.jetbrains.edu.learning.newproject.CourseProjectGenerator
import org.jetbrains.annotations.Nls
import java.io.IOException

/**
 * The main interface provides courses creation for some language.
 *
 * @param Settings container type holds course project settings state
 *
 * If you add any new methods here, please do not forget to add it also to
 * @see com.jetbrains.edu.learning.stepik.hyperskill.HyperskillCourseBuilder
 */
interface EduCourseBuilder<Settings : Any> {
  /**
   * Shows UI for new study item creation
   *
   * @param model some parameters for UI extracted from context where creating action was called
   *
   * @return properties for study item creation
   */
  fun showNewStudyItemUi(
    project: Project,
    course: Course,
    model: NewStudyItemUiModel,
    studyItemCreator: (NewStudyItemInfo) -> Unit
  ) = showNewStudyItemDialog(project, course, model, studyItemCreator = studyItemCreator)

  /**
   * Creates additional content of new study item in project
   * [item] object is already properly initialized and added to corresponding parent.
   */
  fun onStudyItemCreation(project: Project, item: StudyItem) {}

  /**
   * Allows to update project modules and the whole project structure
   */
  fun refreshProject(project: Project, cause: RefreshCause) { }

  fun createInitialLesson(project: Project, course: Course): Lesson? {
    val lessonInfo = NewStudyItemInfo(LESSON + 1, 1, ::Lesson)
    val lesson = CCCreateLesson().createAndInitItem(project, course, course, lessonInfo)
    val taskInfo = NewStudyItemInfo(TASK + 1, 1, ::EduTask)
    val task = CCCreateTask().createAndInitItem(project, course, lesson, taskInfo)
    lesson.addTask(task)
    return lesson
  }

  /**
   * Returns list of templates to create files in new task that supposed to be checked by tests launch.
   *
   * @param info general information about new task
   * @param withSources determines if returning template list should contain task source files (non test files).
   * Can be `false` if parent item is [FrameworkLesson]
   *
   * @see EduTask
   */
  fun getTestTaskTemplates(course: Course, info: NewStudyItemInfo, withSources: Boolean): List<TemplateFileInfo> =
    getDefaultTaskTemplates(course, info, withSources, true)

  /**
   * Returns list of templates to create files in new task that supposed to be run by user.
   * In compiled languages it means that task should have `main` function
   *
   * @param info general information about new task
   * @param withSources determines if returning template list should contain task source files (non test files).
   * Can be `false` if parent item is [FrameworkLesson]
   *
   * @see OutputTask
   * @see ChoiceTask
   * @see TheoryTask
   * @see IdeTask
   */
  fun getExecutableTaskTemplates(course: Course, info: NewStudyItemInfo, withSources: Boolean): List<TemplateFileInfo> =
    getDefaultTaskTemplates(course, info, withSources, false)

  /**
   * Returns default list of templates to create files in new task.
   *
   * If you need to specify different templates for test and executable tasks,
   * override [getTestTaskTemplates] and [getExecutableTaskTemplates] methods respectively.
   *
   * @param info general information about new task
   * @param withSources determines if returning template list should contain source files (non test files).
   * Can be `false` if parent item is [FrameworkLesson]
   * @param withTests determines if returning template list should contain test files
   */
  fun getDefaultTaskTemplates(
    course: Course,
    info: NewStudyItemInfo,
    withSources: Boolean,
    withTests: Boolean
  ): List<TemplateFileInfo> {
    val configurator = course.configurator ?: return emptyList()
    val templates = mutableListOf<TemplateFileInfo>()

    if (withSources) {
      val taskTemplate = if (withTests) taskTemplateName else mainTemplateName
      if (taskTemplate != null) {
        templates += TemplateFileInfo(taskTemplate, joinPaths(configurator.sourceDir, taskTemplate), true)
      }
    }

    if (withTests) {
      val testTemplate = testTemplateName
      if (testTemplate != null) {
        val testFileName = configurator.testFileName.takeIf { it.isNotEmpty() } ?: testTemplate
        templates += TemplateFileInfo(testTemplate, joinPaths(configurator.testDirs.firstOrNull(), testFileName), false)
      }
    }

    return templates
  }

  fun extractInitializationParams(project: Project, info: NewStudyItemInfo): Map<String, String> = emptyMap()

  /**
   * Add initial content for a new task: task and tests files if the corresponding files don't exist.
   * Supposed to use in course creator mode
   *
   * @param task initializing task
   */
  fun initNewTask(project: Project, course: Course, task: Task, info: NewStudyItemInfo, withSources: Boolean) {
    val templates = when (task) {
      is EduTask -> getTestTaskTemplates(course, info, withSources)
      is OutputTask -> {
        val outputTemplate = TemplateFileInfo(
          OUTPUT_PATTERN_NAME,
          joinPaths(course.testDirs.firstOrNull(), OUTPUT_PATTERN_NAME),
          false
        )
        val inputTemplate = TemplateFileInfo(
          INPUT_PATTERN_NAME,
          joinPaths(course.testDirs.firstOrNull(), INPUT_PATTERN_NAME),
          false
        )
        getExecutableTaskTemplates(course, info, withSources) + outputTemplate + inputTemplate
      }
      is ChoiceTask -> getExecutableTaskTemplates(course, info, withSources)
      is TheoryTask -> getExecutableTaskTemplates(course, info, withSources)
      is IdeTask -> getExecutableTaskTemplates(course, info, withSources)
      else -> return
    }

    val params = extractInitializationParams(project, info)

    for (template in templates) {
      val taskFile = template.toTaskFile(params)
      task.addTaskFile(taskFile)
    }
  }

  /**
   * Copy and return text of given `taskFile`.
   * Note: it is needed to override if content of copied text should be modified before copying.
   *
   * @param taskFile - `taskFile` which text will be copied.
   * @param taskDir  - directory of given `taskFile`.
   * @param newTask  - task for which the text is copied.
   * @return return copied text of given [taskFile] or null if can't load it.
   */
  fun getTextForNewTask(taskFile: TaskFile, taskDir: VirtualFile, newTask: Task): String? {
    try {
      val file = EduUtils.findTaskFileInDir(taskFile, taskDir)
      if (file == null) {
        LOG.warn("Can't find a file by path relative to this file for `${taskFile.name}` file")
        return null
      }
      return file.loadEncodedContent()
    }
    catch (e: IOException) {
      LOG.error("Can't load text for `${taskFile.name}` task file", e)
    }
    return null
  }

  val taskTemplateName: String? get() = null
  val mainTemplateName: String? get() = taskTemplateName
  val testTemplateName: String? get() = null

  /**
   * @return object responsible for language settings
   * @see LanguageSettings
   */
  fun getLanguageSettings(): LanguageSettings<Settings>

  fun getSupportedLanguageVersions(): List<String> = emptyList()

  fun getCourseProjectGenerator(course: Course): CourseProjectGenerator<Settings>? = null

  /**
   * Validates a study item name.
   * Returns null if [name] is a correct item name, otherwise returns error message string.
   *
   * @see [com.jetbrains.edu.coursecreator.CCStudyItemPathInputValidator]
   */
  @Nls(capitalization = Nls.Capitalization.Sentence)
  fun validateItemName(project: Project, name: String, itemType: StudyItemType): String? {
    return null
  }

  /**
   * Provides language specific action before study item deletion.
   * Called in write action and only in educator mode
   */
  fun beforeStudyItemDeletion(project: Project, item: StudyItem) {}

  companion object {
    val LOG = Logger.getInstance(EduCourseBuilder::class.java)
  }
}
